File: graph/flow/mcmf.py
from collections import deque
import sys
input = sys.stdin.readline
INF = int(1e12)

class MCMF:
    def __init__(self, n):
        self.n = n
        self.adj = [[] for _ in range(n+1)]
        self.inf = INF
        self.dist, self.visited, self.node, self.edge = [], [], [], []

    def add_edge(self, u, v, capacity, cost):
        self.adj[u].append([v, capacity, cost, len(self.adj[v])])
        self.adj[v].append([u, 0, -cost, len(self.adj[u])-1])

    def spfa(self, s, t):
        self.dist = [self.inf for _ in range((self.n+1))]
        self.visited = [0 for _ in range((self.n+1))]
        self.node = [-1 for _ in range((self.n+1))]
        self.edge = [-1 for _ in range((self.n+1))]
        q = deque([s])
        self.dist[s] = 0
        self.visited[s] = True

        while q:
            u = q.popleft()
            self.visited[u] = False
            for i, (v, cap, cost, _) in enumerate(self.adj[u]):
                if cap > 0 and self.dist[v]>self.dist[u]+cost:
                    self.dist[v] = self.dist[u]+cost
                    self.node[v] = u
                    self.edge[v] = i
                    if not self.visited[v]:
                        q.append(v)
                        self.visited[v] = 1
        return self.dist[t] != self.inf

    def solve(self, s, t, maxf=INF):
        max_flow, min_cost = 0, 0
        while max_flow < maxf and self.spfa(s, t):
            flow = self.inf
            now = t
            while now != s:
                prev = self.node[now]
                edge_idx = self.edge[now]
                flow = min(flow, self.adj[prev][edge_idx][1])
                now = prev
            flow = min(flow, maxf-max_flow)
            max_flow += flow
            min_cost += flow*self.dist[t]
            now = t
            while now != s:
                prev = self.node[now]
                edge_idx = self.edge[now]
                self.adj[prev][edge_idx][1] -= flow
                rev_edge_idx = self.adj[prev][edge_idx][3]
                self.adj[now][rev_edge_idx][1] += flow
                now = prev
        return max_flow, min_cost

mcmf = MCMF(111)

File: graph/flow/mcmf.cpp
#include <bits/stdc++.h>
#define fastio cin.tie(0), cout.tie(0), ios_base::sync_with_stdio(false)
typedef long long ll;
// #define int ll
const int INF = 2e9;
const int MAX_SIZE = 111;
using namespace std;

class MCMF{
public:
    int n;
    vector<vector<tuple<int, int, int, int>>> adj;
    array<int, MAX_SIZE> dist;
    array<bool, MAX_SIZE> inq;
    array<bool, MAX_SIZE> visited;
    array<int, MAX_SIZE> used;

    MCMF(int size): n(size){
        adj.resize(n+1);
    }

    void add_edge(int u, int v, int capacity, int cost){
        adj[u].push_back({v, capacity, cost, (int)adj[v].size()});
        adj[v].push_back({u, 0, -cost, (int)adj[u].size()-1});
    }

    bool spfa(int src, int sink){
        dist.fill(INF); inq.fill(false);
        dist[src] = 0; inq[src] = true;
        queue<int> q; q.push(src);

        while (!q.empty()){
            int now = q.front(); q.pop();
            inq[now] = false;
            for (auto [nxt, cap, cost, dual]: adj[now]){
                if (cap && dist[nxt] > dist[now]+cost){
                    dist[nxt] = dist[now]+cost;
                    if (!inq[nxt]){
                        q.push(nxt);
                        inq[nxt] = true;
                    }
                }
            }
        }
        return dist[sink] < INF;
    }

    int dfs(int src, int sink, int now, int nowflow){
        visited[now] = true;
        if (now == sink) return nowflow;
        for(; used[now]<adj[now].size(); used[now]++){
            auto &[nxt, cap, cost, dual] = adj[now][used[now]];
            if (!visited[nxt] && dist[nxt]==dist[now]+cost && cap){
                int tmp = dfs(src, sink, nxt, min(nowflow, cap));
                if (tmp){
                    cap -= tmp;
                    get<1>(adj[nxt][dual]) += tmp;
                    return tmp;
                }
            }
        }
        return 0;
    }

    pair<int, int> solve(int src, int sink, int maxf=INF){
        int maxflow = 0, mincost = 0;
        while((maxflow < maxf) && spfa(src, sink)){
            used.fill(0);
            int nowflow = 0;
            visited.fill(false);
            while ((nowflow = dfs(src, sink, src, INF))){
                maxflow += nowflow;
                mincost += dist[sink]*nowflow;
                visited.fill(false);
            }
        }
        return {maxflow, mincost};
    }
};

signed main(){
    fastio;
    MCMF mcmf(MAX_SIZE);
}
